README.md
<p align="center">
  <img width="300" src="https://raw.githubusercontent.com/battlefield-portal-community/Image-CDN/main/bf_portal_logo_bc.png#gh-dark-mode-only">
  <img width="300" src="https://raw.githubusercontent.com/battlefield-portal-community/Image-CDN/main/bf_portal_logo_noir.png#gh-light-mode-only">
  </p>

[![Docker Build](https://github.com/battlefield-portal-community/bfportal.gg/actions/workflows/main.yml/badge.svg?branch=main)](https://github.com/battlefield-portal-community/bfportal.gg/actions/workflows/main.yml)

# What is this ?

With the release of battlefield 2042, ripple effect studio, added the ability to make custom game modes called "experiences", currently there is no way to share a game mode outside of the game, this project aims to fill that gap.

# Info

Head over to [bfportal.gg](https://bfportal.gg/), and try it out

## Features

- User system (Discord login for integration with portal community discord server)
- Auto Fill API for forms
- Pagination and website wide search

## How it works

It is pretty straight forward

- You use your discord account to make a new account on the website and submit your experiences.
- On the submission page if u choose to share the playground Url of the experience, the submission form will autofill 😃.

You can later edit your experience if u like.

### Special Thanks to

- [gametools.network](https://gametools.network/) for providing API to autofill submisson form 🥰
- [Wagtail](https://github.com/wagtail/wagtail) for providing an awesome framework that makes managing submissions very easy ✅
- [tailwindcss](https://github.com/tailwindlabs/tailwindcss) for saving the lives of backend developers. 🙏
- [Matavatar](https://discord.com/users/236802771381125120) from BFportal discord community for designing UI of the website 🤝


CHANGELOG.md
#### 2.0.0 (2023-07-23)

##### New Features

* **core:**
  *  add base api url (#202) (39941c24)
  *  add base api url (92cb161a)
  *  Add Legal disclaimers.  (#201) (5caad84b)
  *  add disclaimer for mobile (8e55b91c)
  *  add why i love lawyers (83d2949a)
  *  add Electronic Arts’ Fansite Agreement (f9aa03b2)
  *  migrate to poetry for dependency management (d58aacd8)
  *  upgrade wagtail to 5.0.2 (#185) (38c57736)
  *  upgrade wagtail to 5.0.2 (04e5cb11)
  *  make generation of mock data much better for dev purpose (#183) (39c58893)
  *  remove discriminator from user profile (852ec4b3)
  *  send 404 on invalid profile (c01a9a4a)
  *  allow discord_id that has more than 18 digits (314ae215)
  *  set social auth provider name for mock user (0fc16e1d)
  *  set correct global_name and display_name (f33dcc9a)
  *  improved mock experience creatation; closes #175; closes #176 (38700955)
  *  add management command called  to populate db with mock data (64732972)
  *  add boolean field to profile to mark a user as mock user (5b4b7009)
*  API rework (#199) (ca6fca75)
*  add logging and new page after submit (#181) (4728ba50)
*  use global logger for logging when using runserver (6eed9376)
*  setup logging before starting application (c7839b04)
*  use InterceptHandler to route django logs to loguru; closes #179 (87f8540d)
*  use loguru and logger for gunicorn; closes #180 (1bbd5408)
*  restructure wsgi to make to more modular (ebb12a89)
*  use gunicorn conf to set a few command line args; closes #178 (e62fb1d6)
*  add gunicorn conf (8ebd8ae7)
*  supress wagtail warning and load .env when starting with gunicorn (3e243488)
*  add handling when a user with no profile tries to login on (3608474d)
*  implement new UI for after submit page  (#173) (7ea42237)
*  Create CODE_OF_CONDUCT.md (#172) (508f05e5)
*  check length of token to see if its passed (0d7ca29f)
*  populate env template with default values for local development, update env usage in codebase (3b35021a)
*  update contrib guide (e3562424)
*  update env template (2395eca5)
*  setup docker for local development (48623d71)
*  add broken field to api (#170) (fac31af4)
*  add field  to api (40d42962)
*  remove update-portainer workflow (7a1a18be)
*  add WIP action to manage PRs (2a0c4358)
*  add github sponsors info (7570a546)
* **bfportal.gg:**  Bump version -> 2.0.0 (e6c0c7f1)
* **api:**
  *  add `owner` to api_fields of experience page (2a821728)
  *  add Social accounts to User API response (505899cf)
  *  made api a separate module (98b70306)
  *  serialize  field in ExperiencePage (768cf466)
  *  complete serializers for User Profile (d118c892)
  *  break serializers into modules (4f330c5c)
  *  add Serializer for SubCategory (0c404bb3)
  *  serialize ExperiencesCategory (e016a5a5)
* **tailwind:**
  *  move tailwind config to bfportal folder (#197) (dc2372c3)
  *  move tailwind config to bfportal folder (f316f062)
  *  use standalone tailwind instead of django-tailwind (#196) (7c4b96f4)
  *  add scripts to make dev experience better (e472f7e7)
  *  remove theme app (a2450ff1)
  *  remove django-tailwind (7b378ac8)
* **CI:**
  *  update image to use npm for tailwind (116d66eb)
  *  add caching to docker action (#192) (683264e7)
  *  use github cache to cache image layers (7f7eba6d)
  *  run docker action on all paths (cd9c28ea)
  *  set up buildx in GH actions (40d7e4b3)
  *  set default directory for runner (2963a902)
  *  migrate to a multistage docker build (a7766fe1)
  *  build docker image for main and production branch (#186) (099524c1)
  *  build docker image for main and production branch (50ddd69f)
  *  add default_install_hook_types to pre-commit config to install commit-msg type hooks by default (898324b1)
* **poetry:**  use poetry as the dependency manager (#188) (37b46a51)
* **mock:**
  *  `continue` if a user exists with the same username (e428544e)
  *  add random users as creators; closes #176 (9e10e0e0)
  *  set random states to experience page (119d424b)
  *  remove factory app and merge into one management command called (073bc014)
* **docker:**  add profiles to docker compose for easier development on local (#182) (3edc6540)
* **docs:**  compelete steps to use docker as dev environment (dc7acf51)
* **gh-actions:**
  *  migrate to ghcr.io from docker.pkg.github.com (e4ebc244)
  *  removed dev from push branches in 'on' (e1349849)

##### Chores

* **core:**
  *  add version in package.json (d266e40e)
  *  merge dev into main (#187) (6c9617f2)
*  add `cachetools` to dependencies (018c3385)
*  Merge branch 'dev' into docker-dev-env (46985668)
*  add DB_PASSWORD to template (06d7e001)
*  Merge branch 'dev' into docker-dev-env (86e0ef5e)
*  add  to LOGGING_FORMAT (8f9dd4e9)
*  added loguricorn (2ae1cee1)
*  merge branch 'main' into dev (08e51c58)
*  sort keys into required or not (124a2ab6)
*  Merge branch 'main' into docker-dev-env (f130163b)
*  merge main (f608da84)
*  clean up readme (6a6a1bc4)
*  merge main -> dev (d65df934)
* **lint:**  fix linting errors (e3c93cc3)
* **CI:**
  *  resolve merge conflict (4559ecdf)
  *  fix flake8 and black warnings (11086c57)
* **mock:**  add logging (c9048930)
* **pre-commit:**  bump commit-msg-validator version (7702d6a0)

##### Bug Fixes

* **core:**  remove function (64984b66)
* **CI:**
  *  ignore correct path in collectstatic (4bfa0289)
  *  use raw type tag to get latest tag meta data (#194) (b0e4997e)
  *  use raw type tag to get latest tag meta data (af5fddf6)
  *  set correct flavor in meta-data action (#193) (e76b82f2)
  *  set correct flavor in meta-data action (b7ddaa02)
  *  get metadata for  image in meta-data action (9721dddb)
  *  set correct branch name in docker push (#190) (f5eebb18)
  *  set correct branch name in docker push (bc42640d)
  *  set correct version for meta-data action (#189) (9af686bc)
  *  set correct version for meta-data action (87ce37a8)
*  dependency conflict in draftjs-exporter (f8b5ca46)
*  fix logic when cheking for env lenght 🥲 (78c0f68b)

bfportal/core/views.py
import enum
import os

import requests
from allauth.socialaccount.models import SocialAccount
from core import get_dashboard_url, get_page_url
from core.forms import ExperiencePageForm
from core.models import ExperiencePage, ExperiencesCategory, ExperiencesPage, HomePage
from core.utils.helper import unique_slug_generator
from dal import autocomplete
from django.contrib.auth import get_user_model
from django.contrib.auth.decorators import login_required
from django.http import HttpRequest, HttpResponse
from django.shortcuts import render
from django.utils import timezone
from loguru import logger
from taggit.models import Tag


class EnumMeta(enum.EnumMeta):
    """Meta Class that enables `in` operator for Enums"""

    def __contains__(cls, item):
        try:
            cls(item)
        except ValueError:
            return False
        else:
            return True


class ReportTypes(enum.Enum, metaclass=EnumMeta):
    """All types of report that can be done for an exp"""

    BUGGED = 0
    BROKEN = 1
    XP_FARM = 2
    OK = 4


def send_approve_request(
    page: ExperiencePage, type: str = "new", request: HttpRequest = None
):
    """Tries to send an approval request embed to the

    discord channel specified by APPROVAL_CHANNEL_WEBHOOK_ID env
    """
    if len(token := os.getenv("APPROVAL_CHANNEL_WEBHOOK_TOKEN", "")):
        if len(webhook_id := os.getenv("APPROVAL_CHANNEL_WEBHOOK_ID", "")):
            webhook_url = f"https://discord.com/api/webhooks/{webhook_id}/{token}"
            admin_add = False
            uid = -1
            try:
                uid = SocialAccount.objects.get(user_id=page.owner.id).uid
            except SocialAccount.DoesNotExist:
                admin_add = True

            logger.debug("Trying to send approve request")
            url = get_page_url(request, page)
            if type == "new":
                data = {
                    "content": "New Experience approve request",
                    "embeds": [
                        {
                            "url": url,
                            "title": page.title,
                            "description": page.description[0:200] + ".....",
                            "image": {
                                "url": page.cover_img_url
                                if page.cover_img_url
                                else "https://super-static-assets.s3.amazonaws.com/19d9fbc6-6292-4be8-ac70-5a186b556054%2Fimages%2Fb6495922-b4c7-4002-9c3d-56bfaa5b98b5.jpg"  # noqa: E501
                            },
                            "fields": [
                                {
                                    "name": "Author",
                                    "value": f"<@{uid}>"
                                    if not admin_add
                                    else "**admin**",
                                    "inline": True,
                                },
                                {
                                    "name": "Submitted on",
                                    "value": f"<t:{int(page.first_published_at.timestamp())}>",
                                    "inline": True,
                                },
                                {
                                    "name": "Featured",
                                    "value": ":white_check_mark:"
                                    if page.featured
                                    else ":x:",
                                    "inline": True,
                                },
                                {
                                    "name": "Category",
                                    "value": ":white_small_square: "
                                    + f"\u200B{page.category}",
                                },
                                {
                                    "name": "Sub Categories",
                                    "value": ":white_small_square: "
                                    + "".join(
                                        [f"`{i}` " for i in page.sub_categories.all()]
                                    ),
                                },
                                {
                                    "name": "Tags",
                                    "value": ":white_small_square: "
                                    + "".join([f"`{i}` " for i in page.tags.all()]),
                                },
                            ],
                        }
                    ],
                }
            elif type == "edit":
                data = {
                    "content": "Request to make Changes to a Experience",
                    "embeds": [
                        {
                            "url": url,
                            "title": page.title,
                            "fields": [
                                {
                                    "name": "By",
                                    "value": f"<@{uid}>"
                                    if not admin_add
                                    else "**admin**",
                                }
                            ],
                        }
                    ],
                }
            else:
                data = {
                    "content": "There is Something Wrong with the server pls check 🤐"
                }

            data["token"] = token
            data["components"] = [
                {
                    "type": 1,
                    "components": [
                        {
                            "type": 2,
                            "style": 5,
                            "url": get_dashboard_url(request, page),
                            "label": "Go to Dashboard",
                        }
                    ],
                }
            ]

            headers = {"Content-Type": "application/json"}
            result = requests.post(webhook_url, json=data, headers=headers)
            try:
                result.raise_for_status()
            except requests.exceptions.HTTPError as err:
                logger.debug(
                    f"Error {err} while sending approve request for {page.title}"
                )
                print(result.content)
            else:
                logger.debug(
                    f"{type.capitalize()} Experience request for {page.title} sent successfully "
                )
        else:
            logger.warning(
                "Unable to send Approval request to discord as APPROVAL_CHANNEL_WEBHOOK_ID is not set"
            )
    else:
        logger.warning(
            "Unable to send Approval request to discord as APPROVAL_CHANNEL_WEBHOOK_TOKEN is not set"
        )


@login_required
def submit_experience(request: HttpRequest, home_page: HomePage):
    """Called when a user submits an experience

    This function handles the routing on successful submission
    """
    if request.method == "POST":
        form = ExperiencePageForm(request.POST)
        if form.is_valid():
            logger.debug(f"Saving new exp {form.cleaned_data}")
            new_exp_page: ExperiencePage = form.save(commit=False)
            new_exp_page.slug = unique_slug_generator(new_exp_page)
            new_exp_page.owner = request.user
            new_exp: ExperiencePage = ExperiencesPage.objects.all()[0].add_child(
                instance=new_exp_page
            )
            user_model = get_user_model()
            for pk in form.cleaned_data["creators"]:
                new_exp.creators.add(user_model.objects.get(pk=pk))

            for pk in form.cleaned_data["tags"]:
                new_exp.tags.add(Tag.objects.get(pk=pk))

            for sub_cat in form.cleaned_data["sub_categories"]:
                new_exp.sub_categories.add(sub_cat)

            if new_exp:
                new_exp.unpublish()
                new_exp.save_revision(submitted_for_moderation=True, user=request.user)
                new_exp.first_published_at = timezone.now()
                new_exp.save()
                send_approve_request(new_exp, "new", request)

            return render(
                request, "core/after_submit.html", {"exp_name": new_exp.title}
            )

    else:
        form = ExperiencePageForm()

    return render(
        request,
        "core/submit_experience_page.html",
        {
            "form": form,
            "page": {
                "title": "Submit Experience",
                "description": "Submit your Experience on https://bfportal.gg/",
            },
        },
    )


@login_required
def edit_experience(request: HttpRequest, experience_page: ExperiencePage):
    """Called when a user tries to edit an experience

    handles the routing when a successful edit is done
    """
    if request.method == "POST":
        form = ExperiencePageForm(request.POST, instance=experience_page)
        if form.is_valid():
            logger.debug(form.cleaned_data)
            experience_page.creators.clear()
            experience_page.tags.clear()

            user_model = get_user_model()
            for pk in form.cleaned_data["creators"]:
                experience_page.creators.add(user_model.objects.get(pk=pk))
            for pk in form.cleaned_data["tags"]:
                experience_page.tags.add(Tag.objects.get(pk=pk))

            # logger.debug(form.cleaned_data['tags'])

            experience_page.save_revision(
                submitted_for_moderation=True, user=request.user, changed=True
            )
            send_approve_request(experience_page, "edit", request)
            return render(
                request,
                "core/after_submit.html",
                {"exp_name": experience_page.title, "after_edit": True},
            )
        else:
            return render(
                request,
                "core/submit_experience_page.html",
                {
                    "form": form,
                    "is_edit": True,
                    "disable_preview_embed": True,
                },
            )
    else:
        form = ExperiencePageForm(
            initial={
                "sub_categories": [
                    cat.pk for cat in experience_page.sub_categories.all()
                ],
                "description": experience_page.description,
            },
            instance=experience_page,
        )
        return render(
            request,
            "core/submit_experience_page.html",
            {
                "form": form,
                "is_edit": True,
                "disable_preview_embed": True,
            },
        )


# to sort tags based on how many times its used
# >>> a = ExperiencePageTag.objects.values_list('tag_id').annotate(tag_count=Count('tag_id')).order_by('-tag_count')
# >>> [Tag.objects.get(id=i[0]) for i in a]


def filter_startswith(
    model, starts_with_value: str = None, order_by: str = "id", max_count: int = None
) -> list:
    """Filters objects of a Model

    returns all in DB if no starts_with_value
    else returns the values whose name starts with starts_with_value
    """
    qs = model.objects.all().order_by(order_by)
    if starts_with_value:
        qs = qs.filter(name__istartswith=starts_with_value)
        return qs
    return qs[0:max_count]


class CategoriesAutocomplete(autocomplete.Select2QuerySetView):
    """Used by filter API to get all categories"""

    def get_queryset(self):
        """Returns all the Categories that match the condition"""
        return filter_startswith(ExperiencesCategory, self.q)


class TagsAutocomplete(autocomplete.Select2QuerySetView):
    """Used by filter API to get all tags starting with some value"""

    def get_queryset(self):
        """Returns all the tags that match the condition"""
        return filter_startswith(Tag, self.q, max_count=10)


@login_required
def handle_like_request(request: HttpRequest, page_id):
    """favorite/un-favorite"""
    try:
        page = ExperiencePage.objects.get(id=page_id)
        user = request.user
        user_fav = user.profile.liked
        if page not in user_fav.all():
            user.profile.add_liked_page(page)
        else:
            user.profile.remove_liked_page(page)
        user.save()
        page.save()
        return HttpResponse(f"{page.liked_by.all().count()}", status=201)
    except ExperiencePage.DoesNotExist:
        return HttpResponse("experience does not exits", status=404)


@login_required
def report_experience(request: HttpRequest, report_type: int):
    """View that handles reporting of EXP"""
    if request.method == "POST":
        try:
            report_type = ReportTypes(int(report_type))
            if report_type in ReportTypes:
                if page_id := request.GET.get("id", None):
                    try:
                        exp_page = ExperiencePage.objects.get(id=int(page_id))
                    except ExperiencePage.DoesNotExist:
                        logger.debug(f"Wrong page id {page_id}")
                        return HttpResponse(status=400)
                    save_page = True
                    response_text = "Reported Successfully"
                    match report_type:
                        case ReportTypes.OK:
                            if exp_page.owner == request.user:
                                exp_page.bugged = False
                                exp_page.broken = False
                                exp_page.save_revision(
                                    user=request.user,
                                    submitted_for_moderation=True,
                                    changed=True,
                                )
                                return HttpResponse("Awaiting approval", status=200)
                            else:
                                return HttpResponse(
                                    "Nice try... reported :)", status=404
                                )

                        case ReportTypes.BUGGED:
                            if exp_page.owner == request.user:
                                exp_page.bugged = True
                                response_text = "State Change Successful"
                            else:
                                exp_page.bugged_report.add(request.user)

                        case ReportTypes.BROKEN:
                            if exp_page.owner == request.user:
                                exp_page.broken = True
                                response_text = "State Change Successful"
                            else:
                                exp_page.broken_report.add(request.user)

                        case ReportTypes.XP_FARM:
                            exp_page.xp_farm_report.add(request.user)

                    if save_page:
                        exp_page.save()
                    return HttpResponse(response_text, status=200)
                else:
                    logger.debug(f"Wrong page id {page_id}")
                    return HttpResponse(status=400)
            else:
                raise ValueError()
        except ValueError:
            logger.debug(f"{report_type} not valid report type")
            return HttpResponse(status=404)
    else:
        return HttpResponse(404)


def events_view(request: HttpRequest):
    """Handles /events page"""
    return render(
        request,
        "502.html",
    )
